crates/doublets/src/link.rs

```diff
@@ -0,0 +1,216 @@
+///
+/// This trait is implemented for both regular primitives (usize, u64, etc.)
+/// and NonZero primitives (NonZeroUsize, NonZeroU64, etc.)
+pub trait LinkIndex:
+  Copy + Clone + Eq + PartialEq + Ord + PartialOrd + Debug + Send + Sync + 'static
+{
+  /// The zero value for this type
+  fn zero() -> Self;
+
+  /// Check if the value is zero
+  fn is_zero(&self) -> bool;
+
+  /// Convert from usize
+  fn from_usize(val: usize) -> Self;
+
+  /// Convert to usize
+  fn as_usize(&self) -> usize;
+
+  /// Try to increment the value by one
+  fn checked_add_one(&self) -> Option<Self>;
+
+  /// Try to decrement the value by one
+  fn checked_sub_one(&self) -> Option<Self>;
+}
+
+macro_rules! impl_link_index {
+  ($($t:ty),*) => {
+    $(
+      impl LinkIndex for $t {
+        #[inline]
+        fn zero() -> Self {
+          0
+        }
+
+        #[inline]
+        fn is_zero(&self) -> bool {
+          *self == 0
+        }
+
+        #[inline]
+        fn from_usize(val: usize) -> Self {
+          val as Self
+        }
+
+        #[inline]
+        fn as_usize(&self) -> usize {
+          *self as usize
+        }
+
+        #[inline]
+        fn checked_add_one(&self) -> Option<Self> {
+          self.checked_add(1)
+        }
+
+        #[inline]
+        fn checked_sub_one(&self) -> Option<Self> {
+          self.checked_sub(1)
+        }
+      }
+    )*
+  };
+}
+
+impl_link_index!(
+  usize, u8, u16, u32, u64, u128, isize, i8, i16, i32, i64, i128
+);
+
+macro_rules! impl_link_index_nonzero {
+  ($($nz:ty, $inner:ty),*) => {
+    $(
+      impl LinkIndex for $nz {
+        #[inline]
+        fn zero() -> Self {
+          // NonZero types can't represent zero, so we use 1 as the "zero" sentinel
+          // This is safe because we're treating 1 as the starting index
+          unsafe { Self::new_unchecked(1) }
+        }
+
+        #[inline]
+        fn is_zero(&self) -> bool {
+          self.get() == 1
+        }
+
+        #[inline]
+        fn from_usize(val: usize) -> Self {
+          // Map 0 -> 1, 1 -> 2, etc.
+          unsafe { Self::new_unchecked((val as $inner) + 1) }
+        }
+
+        #[inline]
+        fn as_usize(&self) -> usize {
+          // Map 1 -> 0, 2 -> 1, etc.
+          (self.get() - 1) as usize
+        }
+
+        #[inline]
+        fn checked_add_one(&self) -> Option<Self> {
+          Self::new(self.get().checked_add(1)?)
+        }
+
+        #[inline]
+        fn checked_sub_one(&self) -> Option<Self> {
+          if self.get() <= 1 {
+            None
+          } else {
+            Self::new(self.get() - 1)
+          }
+        }
+      }
+    )*
+  };
+}
+
+impl_link_index_nonzero!(
+  NonZeroUsize,
+  usize,
+  NonZeroU8,
+  u8,
+  NonZeroU16,
+  u16,
+  NonZeroU32,
+  u32,
+  NonZeroU64,
+  u64,
+  NonZeroU128,
+  u128,
+  NonZeroIsize,
+  isize,
+  NonZeroI8,
+  i8,
+  NonZeroI16,
+  i16,
+  NonZeroI32,
+  i32,
+  NonZeroI64,
+  i64,
+  NonZeroI128,
+  i128
+);
```

Comment:
It is better to declare Indexes as just primitives, and decltype a `type Repr = NonZero...` as trees element to reduce memory usage